config/src/sumeragi.rs

```diff
@@ -122,52 +123,8 @@ impl Configuration {
 pub struct TrustedPeers {
     /// Optional list of predefined trusted peers. Must contain unique
     /// entries. Custom deserializer raises error if duplicates found.
-    #[serde(deserialize_with = "deserialize_unique_trusted_peers")]
-    pub peers: HashSet<PeerId>,
-}
-
-/// Custom deserializer that ensures that `trusted_peers` only
-/// contains unique `PeerId`'s.
-///
-/// # Errors
-/// - Peer Ids not unique,
-/// - Not a sequence (array)
-fn deserialize_unique_trusted_peers<'de, D>(deserializer: D) -> Result<HashSet<PeerId>, D::Error>
-where
-    D: serde::Deserializer<'de>,
-{
-    /// Helper, for constructing a unique visitor that errors whenever
-    /// a duplicate entry is found.
-    struct UniqueVisitor(core::marker::PhantomData<fn() -> HashSet<PeerId>>);
-
-    impl<'de> serde::de::Visitor<'de> for UniqueVisitor {
-        type Value = HashSet<PeerId>;
-
-        fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
-            formatter.write_str("a set of unique `Peer::Id`s.")
-        }
-
-        fn visit_seq<S>(self, mut seq: S) -> Result<HashSet<PeerId>, S::Error>
-        where
-            S: serde::de::SeqAccess<'de>,
-        {
-            let mut result = HashSet::new();
-            while let Some(value) = seq.next_element()? {
-                if result.contains(&value) {
-                    return Err(serde::de::Error::custom(format!(
-                        "The peer id: {}'s public key appears twice.",
-                        &value
-                    )));
-                }
-                result.insert(value);
-            }
-
-            Ok(result)
-        }
-    }
-
-    let visitor = UniqueVisitor(core::marker::PhantomData);
-    deserializer.deserialize_seq(visitor)
+    #[serde(deserialize_with = "UniqueVec::deserialize_display")]
```

Comment:
why do you need to refer to deserialization function explicitly? Should it be delegated to the correct function by default?

Reply:
I've found no way to have `deserialize()` using `Display` when it's implemented on `T`, and not using it in other cases.
We need `Display` to have a more precise user-facing error as it was before.

So now default `deserialize()` does not restrict `T` anyhow and provides a message just about a problematic place, `deserialize_debug()` prints also debug repr of duplicated value and `deserialize_display()` prints duplicated value using `Disoplay`.